### PR TITLE
Support complex joins and aggregations in AQE

### PR DESCRIPTION
**Pull request by singcha**
Support complex joins and aggregations in AQE :
- 49420bf - Add query completion event for intermediate stages with no shuffle
- a1d8692 - Fix issue while visiting exchange node. Use separate fragement properties object for child nodes of an exchange node

Test plan - 
Unit tests

== NO RELEASE NOTE ==

### CODE REVIEW

File: presto-spark-base/src/main/java/com/facebook/presto/spark/planner/IterativePlanFragmenter.java [hunk 1]
```diff
@@ -258,7 +258,17 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<FragmentProperti
             }
 
             // don't fragment
-            return context.defaultRewrite(node, context.get());
+            if (node.getScope() == LOCAL) {
```
Comment (rschlussel) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/planner/IterativePlanFragmenter.java:261]: I would merge this above:

if(node.getScope != REMOTE_MATERIALIZED || isFragmentReadyForExecution(node)) {
    super.visitExchange(node, context);
}

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkAdaptiveQueryExecution.java [hunk 2]
```diff
@@ -254,11 +244,11 @@ public Void apply(Try<MapOutputStatistics> result)
                 else {
                     log.info("Fragment %s will not get executed now either because there was no exchange involved (a broadcast is present) or because of an unknown issue.",
                             fragment.getFragment().getId());
+                    publishFragmentCompletionEvent(new FragmentCompletionSuccessEvent(currentFragment.getFragment().getId(), Optional.empty()));
```
Comment (rschlussel) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkAdaptiveQueryExecution.java:247]: why do we publish a fragmentcompletionEvent if no fragment was executed

**Review by rschlussel — COMMENTED** [review]
Might makes sense to separate this into two commits, and then have a description in the body of what problem each is fixing.

Also, commit titles should begin with a capital letter: https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkAdaptiveQueryExecution.java (hunk 2 above)
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkAdaptiveQueryExecution.java:247]: We do that to mark non-shuffle stages as ready for planning. Without this - event polling blocks for success/failure event of subtask which will not be executed if there was no shuffle involved

**Comment by rschlussel** [comment]
Fix the checkstyle errors

**Review by rschlussel — APPROVED** [review]

**Pull request merged**